service/mapreduce_service.cc

```diff
@@ -622,12 +622,161 @@ future<> mapreduce_service::dispatch_to_vnodes(schema_ptr schema, replica::colum
+        auto erm = _cf.get_effective_replication_map();
+        const auto& topo = _mapreducer.get_token_metadata_ptr()->get_topology();
+        auto& tablets = _mapreducer.get_token_metadata_ptr()->tablets().get_tablet_map(_schema->id());
+
+        std::map<locator::tablet_replica, dht::partition_range_vector> ranges_per_tablet_replica_map;
+        for (auto& range : _ranges_left) {
+            auto tablet_id = tablets.get_tablet_id(end_token(range));
+            const auto& tablet_info = tablets.get_tablet_info(tablet_id);
+
+            size_t skipped_non_local_replicas = 0;
+            for (auto& replica : tablet_info.replicas) {
+                if (!db::is_datacenter_local(_req.cl) || topo.get_datacenter(replica.host) == topo.get_datacenter()) {
+                    ranges_per_tablet_replica_map[replica].push_back(range);
+                } else {
+                    ++skipped_non_local_replicas;
+                    if (skipped_non_local_replicas == tablet_info.replicas.size()) {
+                        throw std::runtime_error("No live endpoint available");
+                    }
+                }
+            }
+
+            // can potentially stall e.g. with a large tablet count.
+            co_await coroutine::maybe_yield();
+        }
+
+        _ranges_per_replica = ranges_per_tablet_replica_t(_mapreducer.get_token_metadata_ptr()->get_version(), std::move(ranges_per_tablet_replica_map));
+    }
+
+    std::vector<locator::tablet_replica> get_processing_slots() const {
+        std::vector<locator::tablet_replica> slots;
+        for (const auto& [replica, _] : _ranges_per_replica.get_map()) {
+            for (size_t i = 0; i < _limit_per_replica; ++i) {
+                slots.push_back(replica);
+            }
+        }
+        return slots;
+    }
+
+    future<> dispatch_work_and_wait_to_finish() {
+        while (_ranges_left.size() > 0) {
+            co_await prepare_ranges_per_replica();
+
+            co_await coroutine::parallel_for_each(get_processing_slots(),
+                    [&] (locator::tablet_replica replica) -> future<> {
+                auto& ranges = _ranges_per_replica.get_map().find(replica)->second;
+                for (const auto& range : ranges) {
+                    auto erm = _cf.get_effective_replication_map();
+                    if (!_ranges_per_replica.is_up_to_date(_mapreducer.get_token_metadata_ptr())) {
+                        co_return;
+                    }
+
+                    auto it = _ranges_left.find(range);
+                    if (it != _ranges_left.end()) {
+                        _ranges_left.erase(it);
+                        query::mapreduce_request req_with_modified_pr = _req;
+                        req_with_modified_pr.pr = dht::partition_range_vector{range};
+                        co_await _mapreducer.dispatch_range_and_reduce(erm, _dispatcher, _req, std::move(req_with_modified_pr), replica.host, _result, _tr_state);
```

Comment:
So we're ignoring impact of RF on load when dispatching, and actual replica load will be up to 2*CL. Is that intended?